processor/src/main/java/org/mapstruct/ap/internal/model/GeneratedType.java

```diff
@@ -258,8 +258,10 @@ protected void addIfImportRequired(Collection<Type> collection, Type typeToAdd)
             return;
         }
 
-        if ( needsImportDeclaration( typeToAdd ) ) {
-            collection.add( typeToAdd );
```

Comment:
ok.. how could this work in the first place? Now we don't add typeToAdd itself, but the required imported types..

Reply:
Have a look at the `Type` implementation. We are a bit smart there when we need to do an import and only use the `TopLevel.Nested` for nested. There is a complex logic for computing whether we need to import a type or not.

When I did the initial work for this I changed a lot of places where we were adding types to actually return `Type#getImportedTypes`, but apparently this is not everywhere (considering the bug reports we've been receiving)